### ISSUE TITLE
mmap storage panics if Torrent Dropped when initial piece checks aren't complete

### ISSUE THREAD

**Issue by buaazp** [anacrolix/torrent#211]
Hi @anacrolix ,

When `AddTorrent()`, `queuePieceCheck()` will start many goroutines to `t.verifyPiece()`.
If `t.Drop()` quickly, which close the storage, then the mmap storage will panic because `t.hashPiece` is still doing `io.Copy()`.
I can see the `t.piecesQueuedForHash.Len() > 0` when dropping. 

Besides, I want to do some check to make sure the `queuePieceCheck` is completely done, but I failed. Can you give some `<-chan` or other func to catch this status? 

Codes like below to re-product panic:

```
t, _ := cli.AddTorrent(metaInfo)
t.Drop()
```

This is my panic stacks:

```
unexpected fault address 0x8f2af80
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8f2af80 pc=0x405f4b5]

goroutine 98 [running]:
runtime.throw(0x4781cf9, 0x5)
	/usr/local/opt/go/libexec/src/runtime/panic.go:605 +0x95 fp=0xc420c3ba20 sp=0xc420c3ba00 pc=0x402e0f5
runtime.sigpanic()
	/usr/local/opt/go/libexec/src/runtime/signal_unix.go:374 +0x227 fp=0xc420c3ba70 sp=0xc420c3ba20 pc=0x4044ef7
runtime.memmove(0xc420e38000, 0x8f23000, 0x8000)
	/usr/local/opt/go/libexec/src/runtime/memmove_amd64.s:329 +0x3c5 fp=0xc420c3ba78 sp=0xc420c3ba70 pc=0x405f4b5
github.com/anacrolix/torrent/mmap_span.MMapSpan.ReadAt.func1(0x2328000, 0x4a9d820, 0xc42082c9e0, 0x6759dfac01299f17)
	/Users/zippo/develop/go/src/github.com/anacrolix/torrent/mmap_span/mmap_span.go:45 +0xa1 fp=0xc420c3bab8 sp=0xc420c3ba78 pc=0x43d87f1
github.com/anacrolix/torrent/mmap_span.span.ApplyTo(0xc42087e080, 0x1, 0x1, 0x2328000, 0xc420c3bb30)
	/Users/zippo/develop/go/src/github.com/anacrolix/torrent/mmap_span/span.go:15 +0xb1 fp=0xc420c3bb08 sp=0xc420c3bab8 pc=0x43d8721
github.com/anacrolix/torrent/mmap_span.MMapSpan.ReadAt(0xc42087e080, 0x1, 0x1, 0xc420e38000, 0x8000, 0x8000, 0x2328000, 0x0, 0x8c477edb5ed156c9, 0xbd044f76f7a8fdc6)
	/Users/zippo/develop/go/src/github.com/anacrolix/torrent/mmap_span/mmap_span.go:44 +0x89 fp=0xc420c3bb58 sp=0xc420c3bb08 pc=0x43d8569
github.com/anacrolix/torrent/mmap_span.(*MMapSpan).ReadAt(0xc42089d0c0, 0xc420e38000, 0x8000, 0x8000, 0x2328000, 0xc420013c00, 0x4d6e6c8, 0x0)
	<autogenerated>:1 +0x81 fp=0xc420c3bbb8 sp=0xc420c3bb58 pc=0x43d8f31
io.(*SectionReader).ReadAt(0xc4207fcfc0, 0xc420e38000, 0x8000, 0x8000, 0x28000, 0x86, 0xd8000, 0x86)
	/usr/local/opt/go/libexec/src/io/io.go:504 +0x12e fp=0xc420c3bc08 sp=0xc420c3bbb8 pc=0x40762fe
github.com/anacrolix/torrent/storage.(*mmapStoragePiece).ReadAt(0xc420947e60, 0xc420e38000, 0x8000, 0x8000, 0x28000, 0x81dc6244000c042b, 0xdbb5c1324be53c3d, 0xbd175bdcca440d45)
	<autogenerated>:1 +0x69 fp=0xc420c3bc58 sp=0xc420c3bc08 pc=0x43f2269
github.com/anacrolix/torrent/storage.Piece.ReadAt(0x4aa5480, 0xc420947e60, 0xc4207f4690, 0x23, 0xc420e38000, 0x8000, 0x8000, 0x28000, 0x80, 0x7f80, ...)
	/Users/zippo/develop/go/src/github.com/anacrolix/torrent/storage/wrappers.go:65 +0x15f fp=0xc420c3bce0 sp=0xc420c3bc58 pc=0x43f041f
github.com/anacrolix/torrent/storage.(*Piece).ReadAt(0xc42089d120, 0xc420e38000, 0x8000, 0x8000, 0x28000, 0x8000, 0xc420e38000, 0x0)
	<autogenerated>:1 +0x91 fp=0xc420c3bd48 sp=0xc420c3bce0 pc=0x43f15a1
io.(*SectionReader).Read(0xc4207fcff0, 0xc420e38000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	/usr/local/opt/go/libexec/src/io/io.go:465 +0x7c fp=0xc420c3bd98 sp=0xc420c3bd48 pc=0x407606c
io.copyBuffer(0x5b402a0, 0xc4208b0a10, 0x4a9bf60, 0xc4207fcff0, 0xc420e38000, 0x8000, 0x8000, 0x472e920, 0xc42089d100, 0x5b402a0)
	/usr/local/opt/go/libexec/src/io/io.go:392 +0x123 fp=0xc420c3be08 sp=0xc420c3bd98 pc=0x4075ce3
io.Copy(0x5b402a0, 0xc4208b0a10, 0x4a9bf60, 0xc4207fcff0, 0xc4208b0a10, 0x0, 0x0)
	/usr/local/opt/go/libexec/src/io/io.go:362 +0x68 fp=0xc420c3be68 sp=0xc420c3be08 pc=0x4075a88
github.com/anacrolix/torrent.(*Torrent).hashPiece(0xc4200b1b80, 0x23, 0xc420881300, 0xc42077a400, 0x0)
	/Users/zippo/develop/go/src/github.com/anacrolix/torrent/torrent.go:694 +0x216 fp=0xc420c3bf58 sp=0xc420c3be68 pc=0x449cf56
github.com/anacrolix/torrent.(*Torrent).verifyPiece(0xc4200b1b80, 0x23)
	/Users/zippo/develop/go/src/github.com/anacrolix/torrent/torrent.go:1555 +0x1e4 fp=0xc420c3bfd0 sp=0xc420c3bf58 pc=0x44a2404
runtime.goexit()
	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:2337 +0x1 fp=0xc420c3bfd8 sp=0xc420c3bfd0 pc=0x405e941
created by github.com/anacrolix/torrent.(*Torrent).queuePieceCheck
	/Users/zippo/develop/go/src/github.com/anacrolix/torrent/torrent.go:1588 +0xeb
```

**Comment by anacrolix** [comment]
Thanks for the report. This should be a straightforward fix.

What's your purpose with determining when hashing is done? Is the "Checking" field in https://godoc.org/github.com/anacrolix/torrent#PieceState of any use to you?

**Comment by buaazp** [comment]
Thanks for reply. `PieceState` works. My codes are here below.
But I need an extra `time.Ticker` to check if pieces-hashing is done. Maybe a `<-t.HashDone()` func is preferred.
Or `t.Drop()` is blocked until all pieces hashDone.

```
select {
case <-m.exit:
    return
case <-t.GotInfo():
}
pc := t.NumPieces()

ticker := time.NewTicker(10 * time.Millisecond)
defer ticker.Stop()

for i := 0; ; i++ {
    select {
    case <-m.exit:
        return
    case <-ticker.C:
    }
    var hashing int
    for _, ps := range t.PieceStateRuns() {
        if ps.PieceState.Checking {
            hashing += ps.Length
        }
    }
    log.Printf("[%d] %s hashing: %v/%v", i, t.Name(), hashing, pc)
    if hashing <= 0 {
        break
    }
}
log.Printf("dropping %s", t.Name())
t.Drop()
```

**Comment by anacrolix** [comment]
Let's move this part about getting notification about piece checks into another issue.

**Comment by buaazp** [comment]
OK~ I agree.

**Comment by anacrolix** [comment]
Fixed by d77e939944d3ea6ef50498e19405ee3ff08d846d
